internal/application/services.go

```diff
@@ -0,0 +1,228 @@
+		metav1.ListOptions{
+			LabelSelector: selector,
+		})
+	if err != nil {
+		return result, err
+	}
+
+	// Instead of building a full inverted map from service names to app names here we filter on
+	// the service name to generate just that slice of app names.
+
+	for _, binding := range appBindings.Items {
+		for boundServiceName := range binding.Data {
+			if boundServiceName == serviceName {
+				appName := binding.ObjectMeta.Labels["app.kubernetes.io/name"]
+				result = append(result, appName)
+				break
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// ServicesBoundAppsNames returns a map from the names of services in the specified namespace, to
+// the names of the applications they are bound to. The keys of the map are always a combination of
+// namespace name and service name, to distinguish same-named services in different namespaces (See
+// `ServiceKey` below). The application names never contain namespace information, as they are
+// always in the same namespace as the service referencing them.
+
+func ServicesBoundAppsNames(ctx context.Context, cluster *kubernetes.Cluster, namespace string) (map[string][]string, error) {
+
+	result := map[string][]string{}
+
+	// locate service bindings managed by epinio applications.
+	selector := EpinioApplicationAreaLabel + "=service"
+	selector += ",app.kubernetes.io/component=application"
+	selector += ",app.kubernetes.io/managed-by=epinio"
+
+	appBindings, err := cluster.Kubectl.CoreV1().Secrets(namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: selector,
+		})
+	if err != nil {
+		return result, err
+	}
+
+	for _, binding := range appBindings.Items {
+		appName := binding.ObjectMeta.Labels["app.kubernetes.io/name"]
+		namespace := binding.ObjectMeta.Labels["app.kubernetes.io/part-of"]
+
+		for serviceName := range binding.Data {
+			key := ServiceKey(serviceName, namespace)
+			result[key] = append(result[key], appName)
+		}
+	}
+
+	return result, nil
+}
+
+// ServiceKey constructs a single key string from service and namespace names, for the
+// `servicesToApps` map, when used for services and apps across all namespaces. It uses ASCII NUL
+// (\000) as the separator character. NUL is forbidden to occur in the names themselves. This should
+// make it impossible to construct two different pairs of service/namespace names which map to the
+// same key.
+func ServiceKey(name, namespace string) string {
+	return fmt.Sprintf("%s\000%s", name, namespace)
+}
+
+// DecodeServiceKey splits the given key back into name and namespace.  The name is the first
+// result, the namespace the second.
+func DecodeServiceKey(key string) (string, string) {
+	parts := strings.Split(key, "\0000")
+	return parts[0], parts[1]
+}
+
+// BoundServiceNameSet returns the service names for the services bound to the application by a
+// user, as a map/set
+func BoundServiceNameSet(ctx context.Context, cluster *kubernetes.Cluster, appRef models.AppRef) (NameSet, error) {
```

Comment:
Huh. This function is actually not used. Because the process of binding is sufficiently different from Configurations, where the original code is used. Will remove this thing.

Reply:
Done, commit a03d0ca6272ba0604f9c242ce72162bb6efc7a52